### PR TITLE
EAP7-2012 Activate OIDC in the console

### PR DESCRIPTION
**Pull request by kstekovi**
https://issues.redhat.com/browse/HAL-1832

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -75,6 +75,7 @@ Following is a table of supported environment properties that can be used when r
 | :--------------- | :---------------------------------------------- | :-------------------------------------------------------------------------------------------- |
 | `HAL_IMAGE`      | `quay.io/halconsole/hal-development:latest`     | [HAL standalone image](https://hal.github.io/documentation/get-started/#container) to be used |
 | `WILDFLY_IMAGE`  | `quay.io/halconsole/wildfly-development:latest` | WildFly/JBoss EAP image to be used                                                            |
+| `SSO_IMAGE`      | `quay.io/keycloak/keycloak:latest`              | Keyclaok image to be used as SSO                                                              |
```
Comment (vsalbaba) [inline README.md:78]: Keyclaok/Keycloak

File: packages/testsuite/cypress/e2e/elytron-oidc-client/test-oidc-security-configure-oidc.cy.ts [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    });
+  });
+
+  after(() => {
+    cy.task("stop:containers");
+  });
+
+  beforeEach(() => {
+    // required for "flip" method to see all fields in form with a lots of elements
+    cy.viewport(1000, 2500);
```
Comment (OndrejKotek) [inline packages/testsuite/cypress/e2e/elytron-oidc-client/test-oidc-security-configure-oidc.cy.ts:71]: Can we move this to `before` or is there a reason for setting this before each test?

Comment (kstekovi) [reply packages/testsuite/cypress/e2e/elytron-oidc-client/test-oidc-security-configure-oidc.cy.ts:71]: no, it must be set for every test.

File: packages/testsuite/cypress/e2e/elytron-oidc-client/test-oidc-security-configure-oidc.cy.ts [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+  });
+
+  it("Navigate to elytron OIDC client configuration from home page", () => {
+    cy.navigateTo(wildfly, "home");
+    cy.get("#tlc-configuration").should("be.visible").click();
```
Comment (OndrejKotek) [inline packages/testsuite/cypress/e2e/elytron-oidc-client/test-oidc-security-configure-oidc.cy.ts:76]: I believe we do not need all the `.should("be.visible")` as `click` checks this automatically: https://docs.cypress.io/guides/core-concepts/interacting-with-elements#Scrolling

File: packages/testsuite/cypress/e2e/elytron-oidc-client/test-oidc-security-configure-oidc.cy.ts [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+    cy.get("#subsystems").should("be.visible").click();
+    cy.get("#elytron-oidc-client").should("be.visible").click();
+    cy.get("#elytron-oidc-client > a:nth-child(4)").should("be.visible").click();
```
Comment (OndrejKotek) [inline packages/testsuite/cypress/e2e/elytron-oidc-client/test-oidc-security-configure-oidc.cy.ts:79]: > `a:nth-child(4)`

Why not just `a`? There is just one link.

File: packages/testsuite/cypress/e2e/elytron-oidc-client/test-oidc-security-configure-oidc.cy.ts [hunk 5]
```diff
@@ -0,0 +1,158 @@
+describe("TESTS: Configuring Elytron OIDC client in webconsole", () => {
```
Comment (OndrejKotek) [inline packages/testsuite/cypress/e2e/elytron-oidc-client/test-oidc-security-configure-oidc.cy.ts:1]: > Configuring Elytron OIDC client in webconsole

-> Configuration => Subsystem => Elytron OIDC Client

Comment (OndrejKotek) [reply packages/testsuite/cypress/e2e/elytron-oidc-client/test-oidc-security-configure-oidc.cy.ts:1]: Well, maybe not. But I would expect that this spec would verify it is possible/necessary to log in to Web Console using Keycloak then.

File: packages/testsuite/cypress/e2e/elytron-oidc-client/test-oidc-security-configure-oidc.cy.ts [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+    "access",
+    "authorization",
+    "role-mapping",
+    "Administrator",
+    "include",
+    "user-userwithmappedrole",
+  ];
+  const optionalFields = "a.field-section-toggle-pf";
+  const activationRBACWarning = "#hal-finder-preview > div:nth-child(2) > a";
+  const activationRBACWizadrWarning = "#hal-modal > div > div > div.modal-body > p > p:nth-child(2)";
```
Comment (OndrejKotek) [inline packages/testsuite/cypress/e2e/elytron-oidc-client/test-oidc-security-configure-oidc.cy.ts:19]: > activationRBACWizadrWarning

-> activationRBACWizardWarning

Comment (OndrejKotek) [inline packages/testsuite/cypress/e2e/elytron-oidc-client/test-oidc-security-configure-oidc.cy.ts:19]: > `#hal-modal > div > div > div.modal-body > p > p:nth-child(2)`

Do we really need to be so much precise? Isn't something like `#hal-modal` or `#hal-modal div.modal-body` enough?

File: packages/testsuite/cypress/e2e/elytron-oidc-client/test-oidc-security-configure-oidc.cy.ts [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+    "management",
+    "access",
+    "authorization",
+    "role-mapping",
+    "Administrator",
+    "include",
+    "user-userwithmappedrole",
+  ];
+  const optionalFields = "a.field-section-toggle-pf";
+  const activationRBACWarning = "#hal-finder-preview > div:nth-child(2) > a";
```
Comment (OndrejKotek) [inline packages/testsuite/cypress/e2e/elytron-oidc-client/test-oidc-security-configure-oidc.cy.ts:18]: > `#hal-finder-preview > div:nth-child(2) > a`

What about rather `#hal-finder-preview > div.alert-warning > a`

File: packages/testsuite/cypress/e2e/elytron-oidc-client/test-oidc-security-rbac.cy.ts [hunk 8]
```diff
@@ -33,12 +35,41 @@ describe("TESTS: Access secured by Elytron OIDC client with RBAC", () => {
...
     cy.get("#kc-login").click();
+    cy.get("#tlc-access-control").should("have.text", "Access Control");
     cy.verifyUserName("userwithrole");
     cy.verifyUserRole("Administrator");
     cy.logoutFromWebConsole();
     verifyNotLoggedIn(keycloak);
   });
 
+  it("Logs in successfully and logs out by user with letter case insensiteve role", () => {
+    cy.visit(`/?connect=${wildfly}#home`);
```
Comment (OndrejKotek) [inline packages/testsuite/cypress/e2e/elytron-oidc-client/test-oidc-security-rbac.cy.ts:46]: >    cy.visit(`/?connect=${wildfly}#home`);
>    cy.get("#username").type("userwithrole");
>    cy.get("#password").type("password");
>    cy.get("#kc-login").click();

As those four lines are in every test method, can we introduce a new local method `logIn(username: String, password: String)`instead?

File: packages/testsuite/cypress/support/navigation.ts [hunk 9]
```diff
@@ -46,6 +46,10 @@ Cypress.Commands.add("logoutFromWebConsole", () => {
   cy.get("#logout-link").click();
 });
 
+Cypress.Commands.add("confirmYesInWizard", () => {
+  cy.get('div.modal-footer > button.btn.btn-hal.btn-primary:contains("Yes")').click({ force: true });
```
Comment (OndrejKotek) [inline packages/testsuite/cypress/support/navigation.ts:50]: Why is `force` needed?

File: packages/testsuite/cypress/support/navigation.ts [hunk 10]
```diff
@@ -46,6 +46,10 @@ Cypress.Commands.add("logoutFromWebConsole", () => {
   cy.get("#logout-link").click();
 });
 
+Cypress.Commands.add("confirmYesInWizard", () => {
```
Comment (OndrejKotek) [inline packages/testsuite/cypress/support/navigation.ts:49]: > confirmYesInWizard

-> confirmYesInModal ?

File: packages/testsuite/cypress/support/navigation.ts [hunk 11]
```diff
@@ -116,6 +120,11 @@ declare global {
        * @category Navigation
        */
       logoutFromWebConsole(): void;
+      /**
+       * Confirm resource dialog by click on YES button
```
Comment (OndrejKotek) [inline packages/testsuite/cypress/support/navigation.ts:124]: > resource

-> modal ?

> YES

-> "Yes" ?

File: packages/testsuite/cypress/e2e/elytron-oidc-client/test-oidc-security-configure-oidc.cy.ts (hunk 3 above)
Comment (kstekovi) [reply packages/testsuite/cypress/e2e/elytron-oidc-client/test-oidc-security-configure-oidc.cy.ts]: The visibility verification removed. Yes the Click method is verifying it also.

File: packages/testsuite/cypress/e2e/elytron-oidc-client/test-oidc-security-configure-oidc.cy.ts (hunk 4 above)
Comment (kstekovi) [reply packages/testsuite/cypress/e2e/elytron-oidc-client/test-oidc-security-configure-oidc.cy.ts]: The html selector updated.

File: packages/testsuite/cypress/e2e/elytron-oidc-client/test-oidc-security-configure-oidc.cy.ts [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+  after(() => {
+    cy.task("stop:containers");
+  });
+
+  beforeEach(() => {
+    // required for "flip" method to see all fields in form with a lots of elements
+    cy.viewport(1000, 2500);
+  });
+
+  it("Navigate to elytron OIDC client configuration from home page", () => {
```
[the remark this answers is not in the document]
Comment (kstekovi) [reply packages/testsuite/cypress/e2e/elytron-oidc-client/test-oidc-security-configure-oidc.cy.ts]: updated to "Navigates to elytron OIDC client configuration"

File: packages/testsuite/cypress/e2e/elytron-oidc-client/test-oidc-security-configure-oidc.cy.ts (hunk 6 above)
Comment (kstekovi) [reply packages/testsuite/cypress/e2e/elytron-oidc-client/test-oidc-security-configure-oidc.cy.ts]: Typo fixed.

Comment (kstekovi) [reply packages/testsuite/cypress/e2e/elytron-oidc-client/test-oidc-security-configure-oidc.cy.ts]: No, i also try `#hal-modal > div.modal-body > p > p:nth-child(2)` and it doesn't work.

File: packages/testsuite/cypress/e2e/elytron-oidc-client/test-oidc-security-configure-oidc.cy.ts (hunk 7 above)
Comment (kstekovi) [reply packages/testsuite/cypress/e2e/elytron-oidc-client/test-oidc-security-configure-oidc.cy.ts:18]: No, it didn't work.

File: packages/testsuite/cypress/e2e/elytron-oidc-client/test-oidc-security-rbac.cy.ts (hunk 8 above)
Comment (kstekovi) [reply packages/testsuite/cypress/e2e/elytron-oidc-client/test-oidc-security-rbac.cy.ts]: new local method introduced.

File: packages/testsuite/cypress/support/navigation.ts (hunk 9 above)
Comment (kstekovi) [reply packages/testsuite/cypress/support/navigation.ts]: it isn't required. removed.

File: packages/testsuite/cypress/support/navigation.ts (hunk 10 above)
Comment (kstekovi) [reply packages/testsuite/cypress/support/navigation.ts:49]: we have also method like `confirmAddResourceWizard`, `confirmNextInWizard`, `confirmFinishInWizard`. They are very similar.  The difference is only the text on the button. We could create method like: `confirmModal(buttonText)`.  But this update will trigger changes in a lot of test cases. 

So maybe we can create a issue to remove these methods and replace them general one is a separate MR. WDYT?

File: packages/testsuite/cypress/support/navigation.ts (hunk 11 above)
Comment (kstekovi) [reply packages/testsuite/cypress/support/navigation.ts]: updated.

File: packages/testsuite/cypress/e2e/elytron-oidc-client/test-oidc-security-configure-oidc.cy.ts [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+    cy.get(activationRBACWarning).should("have.text", "Enable RBAC").click();
+    cy.get("#hal-modal-title").should("have.text", "Switch Provider");
+    cy.get(activationRBACWizardWarning).should("contain.text", activationRBACWarningMessage);
+    cy.confirmYesInWizard();
+    cy.verifySuccess();
+    cy.verifyAttribute(wildfly, coreSeviceRbacAddress, "provider", "rbac");
+    cy.verifyAttribute(wildfly, coreSeviceRbacAddress, "use-identity-roles", true);
+  });
+
+  it("Assigne EAP role to user from SSO", () => {
```
Comment (OndrejKotek) [inline packages/testsuite/cypress/e2e/elytron-oidc-client/test-oidc-security-configure-oidc.cy.ts:151]: Assigne -> Assign

File: packages/testsuite/cypress/e2e/elytron-oidc-client/test-oidc-security-configure-oidc.cy.ts [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+    cy.text(roleMapping.id, roleMapping.userName, "userwithmappedrole");
+    cy.text(roleMapping.id, roleMapping.include, "Administrator");
+    cy.get("body > div:nth-child(10)").click();
+    cy.confirmAddResourceWizard();
+    cy.verifySuccess();
+    cy.verifyAttribute(wildfly, mappedRoleAddress, roleMapping.userName, "userwithmappedrole");
+  });
+
+  // this test is dependent on all previous tests. It verify whole configuration together.
+  it("Verify the OIDC configuration", () => {
```
Comment (OndrejKotek) [inline packages/testsuite/cypress/e2e/elytron-oidc-client/test-oidc-security-configure-oidc.cy.ts:165]: > Verify the OIDC configuration

Secure access by Elytron OIDC client

File: packages/testsuite/cypress/support/containers-utils.ts [hunk 15]
```diff
@@ -4,7 +4,7 @@ Cypress.Commands.add("startWildflyContainer", () => {
     {
       name: Cypress.spec.name.replace(/\.cy\.ts/g, "").replace(/-/g, "_"),
       configuration: "standalone-insecure.xml",
-      useNetworkHostMode: false,
+      useNetworkHostMode: true,
```
Comment (OndrejKotek) [inline packages/testsuite/cypress/support/containers-utils.ts:7]: A temporary workaround?

File: packages/testsuite/cypress/e2e/elytron-oidc-client/test-oidc-security-configure-oidc.cy.ts [hunk 16]
```diff
@@ -0,0 +1,181 @@
+describe("TESTS: Configuring Elytron OIDC client in web console", () => {
```
[the remark this answers is not in the document]
Comment (kstekovi) [reply packages/testsuite/cypress/e2e/elytron-oidc-client/test-oidc-security-configure-oidc.cy.ts]: updated, thank you.

File: packages/testsuite/cypress/e2e/elytron-oidc-client/test-oidc-security-configure-oidc.cy.ts (hunk 13 above)
Comment (kstekovi) [reply packages/testsuite/cypress/e2e/elytron-oidc-client/test-oidc-security-configure-oidc.cy.ts]: Typo fixed

File: packages/testsuite/cypress/e2e/elytron-oidc-client/test-oidc-security-configure-oidc.cy.ts (hunk 14 above)
Comment (kstekovi) [reply packages/testsuite/cypress/e2e/elytron-oidc-client/test-oidc-security-configure-oidc.cy.ts]: updated, thank you.

File: packages/testsuite/cypress/support/containers-utils.ts (hunk 15 above)
Comment (kstekovi) [reply packages/testsuite/cypress/support/containers-utils.ts]: yes temporary workaround. new optional argument added.

File: packages/testsuite/cypress/e2e/elytron-oidc-client/test-oidc-security-configure-oidc.cy.ts [hunk 17]
```diff
@@ -0,0 +1,181 @@
...
+  // this test is dependent on all previous tests. It verify whole configuration together.
+  it("Verify the OIDC configuration", () => {
+    // the following two CLI commands are needed to secure server because is used unsecured configuration at start.
+    cy.executeInWildflyContainer(
+      `"/core-service=management/management-interface=http-interface:write-attribute(name=http-authentication-factory,value=management-http-authentication),
+        /core-service=management/management-interface=http-interface:write-attribute(name=http-upgrade,value={sasl-authentication-factory=management-sasl-authentication,enabled=true}),
+        reload"`
+    );
+    cy.visit(`/?connect=${wildfly}#home`);
+    cy.origin("http://localhost:8888", () => {
```
[the remark this answers is not in the document]
Comment (kstekovi) [reply packages/testsuite/cypress/e2e/elytron-oidc-client/test-oidc-security-configure-oidc.cy.ts]: used variable with Keycloak URL address.

**Pull request merged**